src/placeos-frontend-loader/api/remote/remote.cr

```diff
@@ -0,0 +1,16 @@
+module PlaceOS::FrontendLoader
+  abstract class Remote
+    private alias Git = PlaceOS::Compiler::Git
+
+    def initialize(@ref : String, @folder_name : String)
+    end
```

Comment:
I think having `@repository` instead of `@ref` would be clearer here.
Where `@repository` is of the form `<org>/<repo>` and is extracted from the URL passed to the initializer.
Whatever you think is cleanest.

Having an abstract def for the URL base would be good, then you don't need to carry around the full HTTP path.
I actually think there's no such thing as an abstract class def, so you may need to do something like...
```crystal
  def self.host
   {{ raise "self.host unimplemented on #{ @type }" }}
  end 
```

The smart constructer could be implemented along the lines as below,

```crystal
def self.from_repository_url(url : String)
  {% begin %}
    case URL
    {% for remote in @type.subclasses %}
    when /#{ {{remote }}.host }/ then {{ remote }}.new(url)
    {% end %}
    else
      raise "No support for repositories hosted on #{URI.parse(url).host}"
    end
  {% end %}
end
```

Reply:
There may be a better way to do this, you probably don't even need to define the `def self.host` on the abstract parent, the fact it is missing will be enough to make the compiler complain.

Reply:
Disregard the talk about abstract class def stuff, it should be an instance abstract def. Please see my main comment.